samplers/aws/xray/internal/manifest.go

```diff
@@ -0,0 +1,359 @@
+)
+
+const defaultInterval = int64(10)
+
+const manifestTTL = 3600
+
+// Manifest represents a full sampling ruleset and provides
+// option for configuring Logger, Clock and xrayClient.
+type Manifest struct {
+	Rules                          []Rule
+	SamplingTargetsPollingInterval time.Duration
+	refreshedAt                    int64
+	xrayClient                     *xrayClient
+	clientID                       *string
+	logger                         logr.Logger
+	Clock                          util.Clock
+	mu                             sync.RWMutex
+}
+
+// NewManifest return manifest object configured with logging, clock and xrayClient.
+func NewManifest(addr string, logger logr.Logger) (*Manifest, error) {
+	// generate client for getSamplingRules and getSamplingTargets API call
+	client, err := newClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate clientID for sampling statistics
+	clientID, err := generateClientID()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Manifest{
+		xrayClient:                     client,
+		Clock:                          &util.DefaultClock{},
+		logger:                         logger,
+		SamplingTargetsPollingInterval: 10 * time.Second,
+		clientID:                       clientID,
+	}, nil
+}
+
+// Expired returns true if the manifest has not been successfully refreshed in
+// manifestTTL seconds.
+func (m *Manifest) Expired() bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.refreshedAt < m.Clock.Now().Unix()-manifestTTL
+}
+
+// MatchAgainstManifestRules returns a Rule and boolean flag set as true if rule has been match against span attributes, otherwise nil and false
+func (m *Manifest) MatchAgainstManifestRules(parameters sdktrace.SamplingParameters, serviceName string, cloudPlatform string) (*Rule, bool, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	matched := false
+
+	for index := range m.Rules {
+		isRuleMatch, err := m.Rules[index].appliesTo(parameters, serviceName, cloudPlatform)
+		if err != nil {
+			return nil, isRuleMatch, err
+		}
+
+		if isRuleMatch {
+			matched = true
+			return &m.Rules[index], matched, nil
+		}
+	}
+
+	return nil, matched, nil
+}
+
+// RefreshManifestRules writes sampling rule properties to the manifest object.
+func (m *Manifest) RefreshManifestRules(ctx context.Context) (err error) {
+	// get sampling rules from AWS X-Ray console
+	rules, err := m.xrayClient.getSamplingRules(ctx)
+	if err != nil {
+		return err
+	}
+
+	// update the retrieved sampling rules to manifest object
+	m.updateRules(rules)
+
+	return
+}
+
+// RefreshManifestTargets updates sampling targets (statistics) for each rule
+func (m *Manifest) RefreshManifestTargets(ctx context.Context) (err error) {
+	var manifest Manifest
+
+	// deep copy centralized manifest object to temporary manifest to avoid thread safety issue
+	m.mu.RLock()
+	err = copier.CopyWithOption(&manifest, m, copier.Option{IgnoreEmpty: false, DeepCopy: true})
```

Comment:
Yep had a chat with @MrAlias regarding the same. He suggested that we might need to use `reflect` package to build out something like this. I found this open source project which was addressing our need so used it instead writing our own logic. I could not find a good way to do this other than using `reflect` package in go. Just curious, why should we avoid using `reflect` package here? I would appreciate any other better suggestions :)

Reply:
I think hand writing the copying would be fairly straightforward without needing reflect or reflection. I'd generally avoid reflection in production code unless absolutely required to make the code easier to reason about since some times reflection copies can copy in unintended ways causing bugs, but if maintainers are OK with this library than it's fine I guess

Reply:
yep will wait on maintainers feedback on using this library.